libmamba/src/core/virtual_packages.cpp

```diff
@@ -163,6 +163,30 @@ namespace mamba
             return res;
         }
 
+#if (defined(__GNUC__) || defined(__clang__)) && __x86_64__
+        __attribute__((target ("default"))) std::string get_archspec_x86_64()
+        {
+            return "x86_64";
+        }
+        __attribute__((target ("sse4.2"))) std::string get_archspec_x86_64()
+        {
+            return "x86_64_v2";
+        }
+        __attribute__((target ("avx2"))) std::string get_archspec_x86_64()
+        {
+            return "x86_64_v3";
+        }
+        __attribute__((target ("avx512f"))) std::string get_archspec_x86_64()
+        {
+            return "x86_64_v4";
+        }
+#else
+        std::string get_archspec_x86_64()
+        {
+            return "x86_64";
+        }
+#endif
+
```

Comment:
There can be corner cases like Intel's Penryn processors that support SSE4.1 but not POPCNT. Hence, it is probably better to check for more features.
Acc. to
https://en.wikipedia.org/wiki/X86-64#Microarchitecture_levels
and
https://github.com/bminor/glibc/blob/glibc-2.38/sysdeps/x86/get-isa-level.h
the following is not exhaustive but close enough:
```suggestion
        std::string get_archspec_x86_64()
        {
#if (defined(__GNUC__) || defined(__clang__)) && __x86_64__
            /* if (__builtin_cpu_supports ("x86-64-v4")) */
            if (__builtin_cpu_supports("avx512f") && __builtin_cpu_supports("avx512bw")
                && __builtin_cpu_supports("avx512cd") && __builtin_cpu_supports("avx512dq")
                && __builtin_cpu_supports("avx512vl"))
            {
                return "x86_64-v4";
            }
            /* if (__builtin_cpu_supports ("x86-64-v3")) */
            if (__builtin_cpu_supports("avx") && __builtin_cpu_supports("avx2")
                && __builtin_cpu_supports("bmi") && __builtin_cpu_supports("bmi2")
                && __builtin_cpu_supports("fma"))
            {
                return "x86_64-v3";
            }
            /* if (__builtin_cpu_supports ("x86-64-v2")) */
            if (__builtin_cpu_supports("popcnt") && __builtin_cpu_supports("sse3")
                && __builtin_cpu_supports("ssse3") && __builtin_cpu_supports("sse4.1")
                && __builtin_cpu_supports("sse4.2"))
            {
                return "x86_64-v2";
            }
#endif
            return "x86_64";
        }
```
With GCC 12+ and future Clang versions, `__builtin_cpu_supports ("x86-64-v2")` et al., can be called directly.

(Haven't tested this on macOS.)

Reply:
With the function multiversioning code I had, we could also check for multiple CPU features by repeating the target. However using `__builtin_cpu_supports` seems less magic-y than function multi-versioning.

Reply:
> and future Clang versions

Surprised this landed so late, but yeah: https://github.com/llvm/llvm-project/commit/27da15381cbe2ac6fd1319f6409dbbab9a857b7b

Reply:
> With the function multiversioning code I had, we could also check for multiple CPU features by repeating the target. However using `__builtin_cpu_supports` seems less magic-y than function multi-versioning.

Oh, I thought repeating targets would be disjunctive not conjunctive. Maybe I'm just missing something.
In other context I like the multiversioning better, too ;). GCC 12 (Clang not yet) also seems to accept `__attribute__((target ("arch=x86-64-v3")))` which is also nice.

Reply:
GCC and clang internally have a total-ordering of the microarchitecture (unlike archspec which has a partial ordering only) and selects a minimum micro-arch based on the targets given and checks at runtime if the runtime micro-arch is equal or higher.

Reply:
Ah, alright, thanks! :bow: